### ISSUE TITLE
Allow different scorers during (validation) inference

### ISSUE THREAD

**Issue by villmow** [pytorch/fairseq#1828]
## 🚀 Feature Request

During inference one might like to score with different metrics (bleu, rouge, precision/recall, gleu, ...).
Define a common Scorer interface and allow implementing different scorers e.g. `register_eval_scorer()`.

### Motivation

On some datasets I need to score my translation results with ROUGE and Precision/Recall as this is the common practice.

### Pitch

1. Define an `register_eval_scorer` decorator and a protocol that scorers need to follow:
   1.1. A `score(refs, hypos)` method that computes a summable score for a batch and returns that as a dictionary which will be included in `logging_outputs`
   1.2 A `reduce_metrics(logging_outputs, criterion)` that reduces the logging outputs to a single score for the whole validation set and logs this as a scalar or derived scalar to `metrics`.
2. In translation:
   2.1 Remove the BLEU specific arguments from `translation`.
   2.2 Leave and rename inference specific arguments in translation (e.g. `--eval-scorer-generation-args`, `--eval-scorer-remove-bpe`, ...).
   2.3 Generation should only be triggered when an evaluation scorer is set.
   2.4  The references and hypos (the output of the generator) are then passed to a scorer (or multiple scorers) and let the scorer handle the logging of the metrics.

I'm not sure if this could be integrated in `fairseq-generate`, but it seems possible to me.

### Alternatives

An alternative would be to generate with fairseq-generate, parse the output and score it yourself. This is hard to do during validation.

### Additional context

I implemented this for myself using the following code on Translation. This is just a modified version of `--eval-bleu`.

It passes the hypos and references to an `eval_scorer`, which implements a `score(hypos, refs)` method and returns the scores as a dictionary which can be included in `logging_outputs`. The task then calls `reduce_metrics(logging_outputs, criterion)` on that scorer, which should process the logging_outputs and log scalars to metrics.

Note, that this only supports one scorer at a time, while it would be possible to have multiple scorers process the translations. However, I don't know how the `registry` handles multiple arguments.

Modified translation_task.py
```python
from eval_scorer import build_eval_scorer, add_eval_scoring_args

class EvaluatingTranslationTask(TranslationTask):
    @staticmethod
    def add_args(parser):
        """Add task-specific arguments to the parser."""
        TranslationTask.add_args(parser)
        add_eval_scoring_args(parser)

    def build_model(self, args):
        self.eval_scorer = build_eval_scorer(args)
        if self.eval_scorer is not None:
            assert getattr(args, 'eval_scorer_detok', None) is not None, (
                '--eval-scorer-detok is required if using --eval-scorer; '
                'try --eval-scorer-detok=moses (or --eval-scorer-detok=space '
                'to disable detokenization, e.g., when using sentencepiece)'
            )
            detok_args = json.loads(getattr(args, 'eval_scorer_detok_args', '{}') or '{}')
            self.tokenizer = encoders.build_tokenizer(Namespace(
                tokenizer=getattr(args, 'eval_scorer_detok', None),
                **detok_args
            ))

            gen_args = json.loads(getattr(args, 'eval_scorer_args', '{}') or '{}')
            self.sequence_generator = self.build_generator(Namespace(**gen_args))

        return super().build_model(args)

    def valid_step(self, sample, model, criterion):
        loss, sample_size, logging_output = super().valid_step(sample, model, criterion)

        if self.eval_scorer is not None:
            hyps, refs = self._inference(self.sequence_generator, sample, model)

            if self.args.eval_scorer_print_samples:
                logger.info('example hypothesis: ' + hyps[0])
                logger.info('example reference: ' + refs[0])

            scorer_logging_outputs = self.eval_scorer.score(hyps, refs)
            logging_output = {**logging_output, **scorer_logging_outputs}

        return loss, sample_size, logging_output

    def reduce_metrics(self, logging_outputs, criterion):
        """
        added smoothing bleu arguments to sacrebleu
        """

        FairseqTask.reduce_metrics(self, logging_outputs, criterion)

        if self.eval_scorer:
            self.eval_scorer.reduce_metrics(logging_outputs, criterion)

    def _inference(self, generator, sample, model):
        """
        return references and hypos instead of immediate scoring
        """
        def decode(toks, escape_unk=False):
            s = self.tgt_dict.string(
                toks.int().cpu(),
                self.args.eval_scorer_remove_bpe,
                escape_unk=escape_unk,
            )
            if self.tokenizer:
                s = self.tokenizer.decode(s)
            return s

        gen_out = self.inference_step(generator, [model], sample, None)
        hyps, refs = [], []
        for i in range(len(gen_out)):
            hypo = decode(gen_out[i][0]['tokens'])
            ref = decode(
                utils.strip_pad(sample['target'][i], self.tgt_dict.pad()),
                escape_unk=True,  # don't count <unk> as matches to the hypo
            )
            if self.args.eval_scorer_lowercase:
                hypo = hypo.lower()
                ref = ref.lower()
            refs.append(ref)
            hyps.append(hypo)

        return hyps, refs
```

and the corresponding `eval_scorer.py`:

```python
from abc import abstractmethod
import logging
from typing import *

from fairseq import registry, metrics
from fairseq.utils import item
import numpy as np


logger = logging.getLogger(__name__)


build_eval_scorer, register_eval_scorer, EVAL_SCORER_REGISTRY = registry.setup_registry(
    '--eval-scorer',
    default=None,
)

def add_eval_scoring_args(parser):
    parser.add_argument('--eval-scorer-generation-args', type=str, metavar='JSON',
                        help='generation args for validation inference, '
                             'e.g., \'{"beam": 4, "lenpen": 0.6}\'')
    parser.add_argument('--eval-scorer-detok', type=str, default="space",
                        help='detokenizer before computing any scores (e.g., "moses"); '
                             'required if using --eval-bleu; use "space" to '
                             'disable detokenization; see fairseq.data.encoders '
                             'for other options')
    parser.add_argument('--eval-scorer-detok-args', type=str, metavar='JSON',
                        help='args for building the tokenizer, if needed')
    parser.add_argument('--eval-scorer-remove-bpe', nargs='?', const='@@ ', default=None,
                        help='remove BPE before computing any scores')
    parser.add_argument('--eval-scorer-print-samples', action='store_true', default=False,
                        help='print sample generations during validation')
    parser.add_argument('--eval-scorer-lowercase', action='store_true', default=True,
                            help='lowercase outputs/references before computing any score')


class GenerationScorer:
    def __init__(self, args):
        self.args = args

    @abstractmethod
    def score(self, references, hypos) -> Dict[str, Any]:
        """ score references and hypos and output something that fits in logging_outputs"""
        pass

    @abstractmethod
    def reduce_metrics(self, logging_outputs, criterion):
        """reduce multiple score dicts into a single metric"""
        pass


@register_eval_scorer("eval-bleu")
class BleuGenerationScorer(GenerationScorer):
    @staticmethod
    def add_args(parser):
        # options for reporting BLEU during validation
        parser.add_argument('--eval-scorer-bleu-tokenized', action='store_true', default=True,
                            help='if setting, we compute tokenized BLEU instead of sacrebleu')
        parser.add_argument('--eval-scorer-bleu-smooth', choices=['exp', 'floor', 'add-k', 'none'], default='exp',
                            help='smoothing method: exponential decay (default), floor (increment zero counts), add-k (increment num/denom by k for n>1), or none')
        parser.add_argument('--eval-scorer-bleu-smooth-value', type=float, default=0.,
                            help='The value to pass to the smoothing technique, when relevant. Default: %(default)s.')

    def __init__(self, args):
        super().__init__(args)
        self.eval_bleu_order = 4
        try:
            import sacrebleu
        except ImportError:
            raise ImportError("Please install sacrebleu with pip install sacrebleu.")

    def score(self, refs: Iterable[str], hypos: Iterable[str]) -> Dict:
        import sacrebleu
        tokenize = sacrebleu.DEFAULT_TOKENIZER if not self.args.eval_scorer_bleu_tokenized else 'none'
        bleu = sacrebleu.corpus_bleu(
            hypos,
            [refs],
            tokenize=tokenize,
        )

        # create logging output in here
        assert len(bleu.counts) == self.eval_bleu_order
        bleu_logging_output = {
            '_bleu_sys_len': bleu.sys_len,
            '_bleu_ref_len': bleu.ref_len,
        }
        # we split counts into separate entries so that they can be
        # summed efficiently across workers using fast-stat-sync
        for i in range(self.eval_bleu_order):
            bleu_logging_output['_bleu_counts_' + str(i)] = bleu.counts[i]
            bleu_logging_output['_bleu_totals_' + str(i)] = bleu.totals[i]

        return bleu_logging_output

    def reduce_metrics(self, logging_outputs, criterion):
        def sum_logs(key):
            return sum(log.get(key, 0) for log in logging_outputs)

        counts, totals = [], []
        for i in range(self.eval_bleu_order):
            counts.append(sum_logs('_bleu_counts_' + str(i)))
            totals.append(sum_logs('_bleu_totals_' + str(i)))

        if max(totals) > 0:
            # log counts as numpy arrays -- log_scalar will sum them correctly
            metrics.log_scalar('_bleu_counts', np.array(counts))
            metrics.log_scalar('_bleu_totals', np.array(totals))
            metrics.log_scalar('_bleu_sys_len', sum_logs('_bleu_sys_len'))
            metrics.log_scalar('_bleu_ref_len', sum_logs('_bleu_ref_len'))

            def compute_bleu(meters):
                import inspect
                import sacrebleu
                fn_sig = inspect.getfullargspec(sacrebleu.compute_bleu)[0]
                if 'smooth_method' in fn_sig:
                    smooth = {'smooth_method': self.args.eval_scorer_bleu_smooth}
                else:
                    smooth = {'smooth': self.args.eval_scorer_bleu_smooth}

                bleu = sacrebleu.compute_bleu(
                    correct=meters['_bleu_counts'].sum,
                    total=meters['_bleu_totals'].sum,
                    sys_len=meters['_bleu_sys_len'].sum,
                    ref_len=meters['_bleu_ref_len'].sum,
                    smooth_value=self.args.eval_scorer_bleu_smooth_value,
                    **smooth
                )
                return round(bleu.score, 2)

            metrics.log_derived('bleu', compute_bleu)


@register_eval_scorer("eval-precision-recall")
class PrecisionRecallGenerationScorer(GenerationScorer):

    @staticmethod
    def add_args(parser):
        pass

    def __init__(self, args):
        super().__init__(args)


    @staticmethod
    def sentence_score(
            reference: Iterable[Any], hypothesis: Iterable[Any]
    ) -> Tuple[int, int, int, int]:

        tp, fp, tn, fn = 0, 0, 0, 0

        reference_set = set(reference)
        hypothesis_set = set(hypothesis)

        for token in hypothesis:
            if token in reference_set:
                tp += 1
            else:
                fp += 1

        for token in reference:
            if token not in hypothesis_set:
                fn += 1

        return tp, fp, tn, fn
    
    @staticmethod
    def corpus_micro_score(
            refs: Iterable[Iterable[Any]], hypos: Iterable[Iterable[Any]]
    ) -> Tuple[int, int, int, int]:
        tp, fp, tn, fn = 0, 0, 0, 0
        for ref, hypo in zip(refs, hypos):
            sample_tp, sample_fp, sample_tn, sample_fn = PrecisionRecallGenerationScorer.sentence_score(ref, hypo)
            tp += sample_tp
            fp += sample_fp
            tn += sample_tn
            fn += sample_fn
        return tp, fp, tn, fn

    def score(self, refs, hypos) -> Dict:

        assert len(refs) == len(hypos)
        total = len(refs)
        # tokenize words FIXME maybe use some tokenizer
        refs = (ref.split() for ref in refs)
        hypos = (hypo.split() for hypo in hypos)

        tp, fp, tn, fn = self.corpus_micro_score(refs, hypos)

        precision_logging_output = {
            '_pr_tp': tp,
            '_pr_fp': fp,
            '_pr_tn': tn,
            '_pr_fn': fn,
            '_pr_total': total
        }

        return precision_logging_output

    def reduce_metrics(self, logging_outputs, criterion):
        def sum_logs(key):
            return sum(log.get(key, 0) for log in logging_outputs)

        total = sum_logs('_pr_total')
        if total:
            metrics.log_scalar('_pr_tp', sum_logs('_pr_tp'))
            metrics.log_scalar('_pr_fp', sum_logs('_pr_fp'))
            metrics.log_scalar('_pr_fn', sum_logs('_pr_fn'))

            def compute_precision(meters):
                _precision = precision(
                    tp=meters['_pr_tp'].sum,
                    fp=meters['_pr_fp'].sum,
                )
                return round(item(_precision * 100), 2)

            def compute_recall(meters):
                _recall = recall(
                    tp=meters['_pr_tp'].sum,
                    fn=meters['_pr_fn'].sum,
                )
                return round(item(_recall * 100), 2)

            def compute_f_measure(meters):
                _f_measure = f_measure(
                    tp=meters['_pr_tp'].sum,
                    fp=meters['_pr_fp'].sum,
                    fn=meters['_pr_fn'].sum,
                )

                return round(item(_f_measure * 100), 2)

            metrics.log_derived('recall', compute_recall)
            metrics.log_derived('precision', compute_precision)
            metrics.log_derived('f-1', compute_f_measure)


def precision(tp: int, fp: int) -> float:
    return tp / (tp + fp) if (tp + fp) > 0 else 0.


def recall(tp: int, fn: int) -> float:
    if (tp + fn) == 0:
        return 0.
    return float(tp) / (tp + fn)


def f_measure(tp: int, fp: int, fn: int, beta=1) -> float:
    f_precision = precision(tp, fp)
    f_recall = recall(tp, fn)
    if f_precision == 0 or recall == 0:
        return 0.
    else:
        return (1 + beta**2) * (f_precision * f_recall) / ((beta**2 * f_precision) + f_recall)
```

**Comment by myleott** [comment]
Great idea! Would you mind submitting a PR with this?

**Comment by villmow** [comment]
Sure :)

**Comment by AlanHassen** [comment]
`        for token in hypothesis:
            if token in reference_set:
                tp += 1
            else:
                fp += 1

        for token in reference:
            if token not in hypothesis_set:
                fn += 1
`

I'm rather unsure if we should calculate the tp based on a set operation. Isn't the position the token has also relevant? In the end, it's a true positive, if our model can predict the correct token for a given position.

**Comment by villmow** [comment]
I think this depends on the definition of the scoring. I compute the overlap of unigrams (each word) between the predictions and references, which is the same as ROUGE-1. This may not be a very good metric for longer sequences (and maybe for sequences at all), but it is used in my task.

In this case a `true positive` would be any word from the *hypothesis* that appears in the *reference* regardless of the position. I use sets only to check the membership in either reference or hypothesis, as member lookups in sets are O(1) instead of O(n) for lists. 

I'm not even sure if this Scorer is wanted, as it basically is a stripped down ROUGE-1 implementation. I just left it here for completeness. One could implement a full ROUGE scorer, with ROUGE-1, ROUGE-2, ROUGE-L instead.

Btw., I have prepared the Pull Request just need to run the translation example to see if everything works, but I'm currently running out of GPUs. So it might take a little.
